### PR TITLE
chore(buffers): Make vector-core buffers generic

### PR DESCRIPTION
**Pull request by blt**
This pull request makes the buffers code generic over a type `T`. Integration with `Event` is maintained by defining `TryFrom<Bytes>` and `TryInto<Bytes>` implementations for that type. Getting the exact trait bounds right took a while, so if you see a place where constraints can be shaved down please do call it out. 

There has been some code re-oganization inside the buffers sub-tree but I have left it all bundled into core's sole package. Once this PR lands #7407 is much less traumatic to achieve.

### CODE REVIEW

File: lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    <T as TryFrom<bytes::Bytes>>::Error: Debug,
+{
+}
+unsafe impl<T> Sync for Writer<T>
+where
+    T: Send + Sync + Unpin + TryInto<Bytes> + TryFrom<Bytes>,
+    <T as TryInto<bytes::Bytes>>::Error: Debug,
+    <T as TryFrom<bytes::Bytes>>::Error: Debug,
+{
+}
```
Comment (ktff) [inline lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs:50]: How come we need `Sync` now? Was it auto implemented before?

File: lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    pub(crate) write_notifier: Arc<AtomicWaker>,
+    pub(crate) blocked_write_tasks: Arc<Mutex<Vec<Waker>>>,
+    pub(crate) writebatch: Writebatch<Key>,
+    pub(crate) batch_size: usize,
+    pub(crate) max_size: usize,
+    pub(crate) current_size: Arc<AtomicUsize>,
+    pub(crate) slot: Option<T>,
+}
+
+// Writebatch isn't Send + Send, but the leveldb docs explicitly say that it's
```
Comment (ktff) [inline lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs:35]:
```suggestion
// Writebatch isn't Send + Sync, but the leveldb docs explicitly say that it's
```

File: lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        data_dir: path.parent().expect("always a parent"),
+    })?;
+    Ok(db.value_iter(ReadOptions::new()).map(|v| v.len()).sum())
+}
+
+impl<T> Buffer<T>
+where
+    T: Send + Sync + Unpin + TryInto<Bytes> + TryFrom<Bytes>,
+    <T as TryInto<bytes::Bytes>>::Error: Debug,
+    <T as std::convert::TryFrom<bytes::Bytes>>::Error: Debug,
```
Comment (ktff) [inline lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs:65]:
```suggestion
    <T as std::convert::TryFrom<bytes::Bytes>>::Error: Debug + Display,
```

File: lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs (hunk 1 above)
Comment (blt) [reply lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs:50]: It was auto-implemented before. I don't love needing to declare `Sync` directly and I would like to see if there's some tweak to the type logic that'll get us back to auto-implemented. I'll leave this comment open as a TODO.

File: lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs (hunk 3 above)
Comment (blt) [reply lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs:65]: `Display` wasn't needed here to compile and when feasible I prefer to leave type constraints lax.

**Comment by blt** [comment]
Out of a sense of paranoia around the new serialization work I've introduce quickcheck testing for serialize -> deserialize being the identity function in b57927a. Amusingly this has discovered a quirk in prost: it doesn't preserve the sign of floating point zero. Consider that my test failures are all turning up failures where we generate a `-0.0` but once the type goes through prost ser/de it has `0.0`.

**Comment by blt** [comment]
> Out of a sense of paranoia around the new serialization work I've introduce quickcheck testing for serialize -> deserialize being the identity function in [b57927a](https://github.com/timberio/vector/commit/b57927af7e04676ec9f57e68b73bf4fa6b04a0c5). Amusingly this has discovered a quirk in prost: it doesn't preserve the sign of floating point zero. Consider that my test failures are all turning up failures where we generate a `-0.0` but once the type goes through prost ser/de it has `0.0`.
> 
> I'm not sure if this is an issue with protobuf the specification or prost. It does seem like, for vector, a zero is a zero whether positive or negative.

Hmm, actually a quick experiment in a clean project suggests that prost does preserve the sign of zero, so the problem is in our project.

**Comment by blt** [comment]
> Out of a sense of paranoia around the new serialization work I've introduce quickcheck testing for serialize -> deserialize being the identity function in [b57927a](https://github.com/timberio/vector/commit/b57927af7e04676ec9f57e68b73bf4fa6b04a0c5). Amusingly this has discovered a quirk in prost: it doesn't preserve the sign of floating point zero. Consider that my test failures are all turning up failures where we generate a `-0.0` but once the type goes through prost ser/de it has `0.0`.
> 
> I'm not sure if this is an issue with protobuf the specification or prost. It does seem like, for vector, a zero is a zero whether positive or negative.

For instance: 

```
                     Quantile {
                         upper_limit: -198912.0,
<                        value: -0.0,
>                        value: 0.0,
                     },
                     Quantile {
                         upper_limit: -47488.0,
                         value: 219648.0,
                     },
                     Quantile {
                         upper_limit: -359936.0,
<                        value: -0.0,
>                        value: 0.0,
                     },
```

File: lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs (hunk 3 above)
Comment (ktff) [reply lib/vector-core/src/buffers/disk/leveldb_buffer/mod.rs]: Yes, but `to leave type constraints lax` fair enough. Let it stay like that.

**Comment by ktff** [comment]
> for vector, a zero is a zero whether positive or negative.

Yep, I don't think we need to have a special case/correct for this so this protobuf/prost quirk seems fine to me. It's comparable how, in some places, we turn non utf8 strings into utf8 by adding special chars.

**Comment by blt** [comment]
> > Out of a sense of paranoia around the new serialization work I've introduce quickcheck testing for serialize -> deserialize being the identity function in [b57927a](https://github.com/timberio/vector/commit/b57927af7e04676ec9f57e68b73bf4fa6b04a0c5). Amusingly this has discovered a quirk in prost: it doesn't preserve the sign of floating point zero. Consider that my test failures are all turning up failures where we generate a `-0.0` but once the type goes through prost ser/de it has `0.0`.
> > I'm not sure if this is an issue with protobuf the specification or prost. It does seem like, for vector, a zero is a zero whether positive or negative.
> 
> For instance:
> 
> ```
>                      Quantile {
>                          upper_limit: -198912.0,
> <                        value: -0.0,
> >                        value: 0.0,
>                      },
>                      Quantile {
>                          upper_limit: -47488.0,
>                          value: 219648.0,
>                      },
>                      Quantile {
>                          upper_limit: -359936.0,
> <                        value: -0.0,
> >                        value: 0.0,
>                      },
> ```

This addressed in 3c9d35c.

File: lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs (hunk 1 above)
Comment (blt) [reply lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs]: Good call. Removed in 7d2501f

File: lib/vector-core/src/bytes.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+//! This module defines traits that allow conversion two and from `bytes`
```
Comment (tobz) [inline lib/vector-core/src/bytes.rs:1]: `two` -> `to`

Comment (blt) [reply lib/vector-core/src/bytes.rs:1]: Oops

Comment (blt) [reply lib/vector-core/src/bytes.rs]: Corrected in 8750f08.

File: lib/vector-core/src/event/test/serialization.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        let mut buffer = BytesMut::with_capacity(64);
+        {
+            let res = Event::encode(event, &mut buffer);
+            assert!(res.is_ok());
+        }
+        {
+            let res = Event::decode(buffer);
+            let actual: Event = res.unwrap();
+            // While Event does implement PartialEq we prefer to use PartialOrd
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/src/event/test/serialization.rs:24]: I've expanded the comment in 8750f08. How's that strike you?

[the remark this answers is not in the document]
Comment (tobz) [reply lib/vector-core/src/event/test/serialization.rs:24]: I still (personally) don't grok the difference here but the comment seems detailed enough to explain this idiosyncrasy. 👍🏻

[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/src/event/test/serialization.rs:24]: It's aggressively goofy.

**Review by lukesteensen — APPROVED** [review]
Looks good to me! Appreciate the extra confidence from quickcheck. That'll also come in handy when we (hopefully) simplify a bit of the `Event -> EventProto -> Bytes` path.

**Pull request merged**